rust/libnewsboat/src/history.rs

```diff
@@ -0,0 +1,142 @@
+use std::fs::OpenOptions;
+use std::io::{self, BufRead, BufReader, BufWriter, Write};
+use std::path::Path;
+
+pub struct History {
+    idx: usize,
+    lines: Vec<String>,
+}
+
+impl History {
+    pub fn new() -> History {
+        History {
+            idx: 0,
+            lines: Vec::new(),
+        }
+    }
+    pub fn add_line(&mut self, line: String) {
+        if !line.is_empty() {
+            self.lines.insert(0, line)
+        }
+        self.idx = 0;
+    }
+    pub fn next(&mut self) -> String {
+        match self.idx {
+            0 => String::from(""),
+            _ => {
+                self.idx -= 1;
+                self.lines[self.idx].to_string()
```

Comment:
Can't a more explicit `clone()` be used here? It's a bit confusing to see `to_string` when I know that `lines` is a `Vec` of `String` already. `clone()` would nicely explain that what we really doing is copying a string. Ditto in `prev()`, below.